packages/darklang/languageTools/nameResolver.dark

```diff
@@ -0,0 +1,280 @@
+            )
+
+      let namesToTry
+        (given: List<String>)
+        (modules: List<String>)
+        : List<List<String>> =
+        match modules with
+        | [] -> [ given ]
+        | _ ->
+          let rest = Stdlib.List.dropLast modules
+          let current = Stdlib.List.append modules given
+          Stdlib.List.append [ current ] (namesToTry given rest)
+
+      let resolve
+        (constructor: String -> 'name)
+        (parser: String -> Stdlib.Result.Result<String>)
+        (nameErrorType: LanguageTools.RuntimeErrors.NameResolution.NameType)
+        (packageThingExists: String -> Bool)
+        (builtinThingExists: String -> Bool)
+        (allowError: Bool)
+        (currentModule: List<String>)
+        (name: LanguageTools.WrittenTypes.Name)
+        : LanguageTools.ProgramTypes.NameResolution<LanguageTools.ProgramTypes.FQName.FQName<'name>> =
+
+        match name with
+        | Unresolved given ->
+          let result =
+            Stdlib.List.fold
+              (NameResolver.namesToTry given currentModule)
+
+              (Stdlib.Result.Result.Error(
+                LanguageTools.RuntimeErrors.NameResolution.Error
+                  { nameType = nameErrorType
+                    errorType =
+                      LanguageTools.RuntimeErrors.NameResolution.ErrorType.NotFound
+                    names = given }
+              ))
+
+              (fun currentResult pathToTry ->
+                match currentResult with
+                | Ok _ -> currentResult
+                | Error _ ->
+                  let newResult =
+                    resolveName
+                      constructor
+                      parser
+                      pathToTry
+                      packageThingExists
+                      builtinThingExists
+                      nameErrorType
+
+                  match newResult with
+                  | Ok _ -> newResult
+                  | Error _ -> currentResult)
+
+          match result with
+          | Ok result -> Stdlib.Result.Result.Ok result
+          | Error err ->
+            if allowError then
+              Stdlib.Result.Result.Error err
+            else
+              Stdlib.Result.Result.Error "TODO: Raise exception"
+
+
+        | _ -> Stdlib.Result.Result.Error "TODO : builtin and user program"
+
+
+
+      module TypeName =
+        let packageTypeExists (typeName: String) : Bool =
+
+          let typ =
+            Stdlib.HttpClient.request
+              "GET"
+              ("http://dark-packages.dlio.localhost:11003/type/by-name/" ++ typeName)
+              []
+              []
+
+          let statusCode =
+            typ
+            |> Stdlib.Result.map (fun response -> response.statusCode)
+            |> Builtin.unwrap
+
+          match statusCode with
+          | 200L -> true
+          | e -> false
+
+        let builtinThingExists (typeName: String) : Bool =
+          // TODO: check if builtin type exists
+          true
```

Comment:
Maybe `false` and replace the TODO with "we don't currently have any Builtin types, and don't expect to)